src/react-integration/hooks/useResource.ts

```diff
@@ -1,49 +1,36 @@
-import { ReadShape, Schema, RequestResource, SchemaOf } from '~/resource';
+import { ReadShape, Schema, SchemaOf } from '~/resource';
 import useCache from './useCache';
 import useRetrieve from './useRetrieve';
 import useError from './useError';
 import { useMemo } from 'react';
 
+import hasUsableData from './hasUsableData';
+
 type ResourceArgs<
   S extends Schema,
   Params extends Readonly<object>,
   Body extends Readonly<object | string> | void
 > = [ReadShape<S, Params, Body>, Params | null];
 
-/** If the invalidIfStale option is set we suspend if resource has expired */
-function hasUsableData<
-  S extends Schema,
-  Params extends Readonly<object>,
-  Body extends Readonly<object | string> | void
->(
-  resource: RequestResource<ReadShape<S, Params, Body>> | null,
-  fetchShape: ReadShape<S, Params, Body>,
-) {
-  return !(
-    (fetchShape.options && fetchShape.options.invalidIfStale) ||
-    !resource
-  );
-}
-
 /** single form resource */
 function useOneResource<
   Params extends Readonly<object>,
   Body extends Readonly<object | string> | void,
   S extends Schema
->(fetchShape: ReadShape<S, Params, Body>, params: Params | null) {
+>(
+  fetchShape: ReadShape<S, Params, Body>,
+  params: Params | null,
+): typeof params extends null ? null : NonNullable<typeof resource> {
+  // maybePromise is undefined when data is stale or params is null
   const maybePromise = useRetrieve(fetchShape, params);
+  // resource is null when it is not in cache or params is null
   const resource = useCache(fetchShape, params);
-
-  if (
-    !hasUsableData(resource, fetchShape) &&
-    maybePromise &&
-    typeof maybePromise.then === 'function'
-  )
-    throw maybePromise;
   const error = useError(fetchShape, params, resource);
+
+  if (!hasUsableData(resource, fetchShape) && maybePromise) throw maybePromise;
   if (error) throw error;
 
-  return resource as NonNullable<typeof resource>;
+  return resource as any;
```

Comment:
seems like a type regression, why did you do this?

Reply:
basically i need to force typescript to say resource is not null. but this is only in the case params is not null. anyway, this is least verbose way to do it because i can just put the actual return type in the function def.